### ISSUE TITLE
Productionize dump-ledger (a proper API for staking/delegating/supercharging info)

### ISSUE THREAD

**Issue by bkase** [MinaProtocol/mina#7737]
## Summary

The easiest path towards making staking/delegating and supercharging info easily available is to productionize the dump-ledger utility and ask that folks store a copy of this every time it changes. Luckily, since we can always dump current and next epoch ledgers from a synced node, we should only require uploads once every 28 days.

Similar to the other archiving redundancy systems, we'll also want to create a process for recovering this data should you miss the upload period without needing to trust the person giving your the dumped ledger.

## Specific Tasks

1. Add the ability to dump specific ledgers. We should have "current epoch ledger", "next epoch ledger" to service the staking requirements. Additionally, include the "current staged ledger".
2. Add a command for recomputing the merkle root and total currency of a dumped ledger. This can be used by people to make sure it matches the epoch ledger in the consensus state for a particular block they want to get the staking information for. We can also use this to verify that information given to us by an untrusted source is correct.
3. Improve usability: 
* Co-locate all commands related to ledger dumping and manipulation described here under a `coda ledger` subcommand; eg. `coda ledger export currentEpochLedger`, `coda ledger hash foo.json`, `coda ledger export stagedLedger`. The details of this can be determined by the implementer.
* Make JSON exporting the default.
* Ensure there are good error messages.
* Nice to have: (only do this if it's very simple after doing the other work) Connect this to GraphQL too -- maybe it will replace #4825 
4. Write documentation: Include information on "archive redundancy" about uploading this data at least every 28 days and recovery from a backup. Include information on a soon-to-exist "staking service guidelines" page on how to use this information to determine stake from delegators and superchargability of nodes.
5. Make sure we upload this information to the cloud in our infrastructure. Using a similar google cloud storage flow as we do for other data seems sufficient here. We can run this on a cron. Maybe we do this once per day. Ideally we do this on a node that doesn't produce blocks.
6. Determine if we actually need to blit the rocksdb DB and THEN dump the JSON out of process

**Comment by garethtdavies** [comment]
As some additional background, I added a few fields to the dumped staking ledger and exposed via a GraphQL API for use in developing some staking pool tooling. What I found most useful was:

* Ability to filter the query by delegate
* Adding the timed/untimed calculation for a delegate so you know if it will be a supercharged reward. In the data I am testing with all accounts are now untimed so `timed_in_epoch` is false but I added an `untimed_slot` as I guess it is possible to become untimed during the epoch.

As an example the following request shows that there are no timed balances in the delegations to the following address:

```
{
  stakes(
    query: {delegate: "B62qrKMAUagWp1VJhvb1AA6y1G4P4tr3inYyPqDRq1VyprA1wmVthcV"}
  ) {
    public_key
    balance
    timing {
      cliff_amount
      cliff_time
      initial_minimum_balance
      vesting_increment
      vesting_period
      untimed_slot
      timed_in_epoch
    }
  }
}
```

returns:

```
{
  "data": {
    "stakes": [
      {
        "balance": 65949.4,
        "public_key": "B62qrAoKfgQx8oXMVQauRs9WYzXpCmYqQrLxivgc8qMYVkK9eH9GjjF",
        "timing": null
      },
      {
        "balance": 71771.5149999,
        "public_key": "B62qrKMAUagWp1VJhvb1AA6y1G4P4tr3inYyPqDRq1VyprA1wmVthcV",
        "timing": null
      },
      {
        "balance": 65958.8001,
        "public_key": "B62qoDaiV18ZGhLfb3j2SuDQEhjyM7c8sDppmvpAyE3cXjutPtHLBfA",
        "timing": {
          "cliff_amount": 12000,
          "cliff_time": 150,
          "initial_minimum_balance": 60000,
          "timed_epoch_end": null,
          "timed_in_epoch": false,
          "untimed_slot": 2070,
          "vesting_increment": 150,
          "vesting_period": 6
        }
      }
    ]
  }
}
```

Available via:

```
curl -d '{"query":"{\n  stakes(\n    query: {delegate: \"B62qrKMAUagWp1VJhvb1AA6y1G4P4tr3inYyPqDRq1VyprA1wmVthcV\"}\n  ) {\n    public_key\n    balance\n    timing {\n      cliff_amount\n      cliff_time\n      initial_minimum_balance\n      vesting_increment\n      vesting_period\n      untimed_slot\n      timed_in_epoch\n      timed_epoch_end\n    }\n  }\n}\n"}' -H 'Content-Type: application/json' https://graphql.minaexplorer.com
```

**Comment by bkase** [comment]
@garethtdavies and others -- if we don't fix the existing endpoint for mainnet, but provide an example archive node query to get all this information for any arbitrary epoch in the past (so strictly more powerful than this graphql endpoint was able to do)

**Comment by bkase** [comment]
cc @garethtdavies -- latest design is reflected in the body of the issue. Unfortunately, there are some nuances in getting the data into the archive node properly.

**Comment by psteckler** [comment]
For item (6), the `ledger export` subcommand uses RPC to grab the entire ledger, as a list of accounts, from the daemon. So the ledger is read from the db before sending it over RPC.

We could add something like `ledger export-local`, that connects directly to the db, and reads and converts to JSON entries lazily. The exported ledger wouldn't contain anything in masks or any other uncommitted entries. We could also make a separate app for this purpose.

**Comment by gaia** [comment]
https://docs.minaexplorer.com/minaexplorer/calculating-payments
states `We will validate this on the testnet and collaborate on a script for all pool operators to use (waiting on this issue)`
but the issue has been closed: https://github.com/MinaProtocol/mina/issues/7737
is this tool available now?

Also, @bkase#2492 & @garethtdavies#4963, Tezos has a very mature tool for calculating payments to delegations (the standard, every validator uses it). While the protocol and actual calculations obviously differ, it might be a good a idea to take a look at some of the options in this tool. Worth mentioning:

a) you can set different rates for different delegators (to for example, calculate with 0% fee the service for partners in the validator service, lower fees for larger delegators, etc. the use cases are plenty)
b) one does not need to keep an archive node, there are three back ends for calculations: Tezos RPC, tzstats API and TzKT API. Naturally these APIs are fed by an archival node, but they allow for higher level queries of the chain, which is useful for developers looking to build on top of the chain.

I can also send you some sample output of how this tool runs if you find this would be useful.
